### DISCUSSION TITLE
How to zip two IncrementalValuesProviders?

### DISCUSSION THREAD

**Discussion by n9** [dotnet/roslyn#74674] [Q&A]
Currently, there is `IncrementalValueProviderExtensions.Combine` that combines either two `IncrementalValueProvider`s or one `IncrementalValuesProvider` with one `IncrementalValueProvider`. I need to zip two `IncrementalValuesProvider`s into a new one by a key.

#### My use case:

I need to generate a partial class from two sources:
- a user-provided interface (which the generated partial class will implement)
- a user-provided partial class (which the generator will also use while generating the generated partial class)

#### Result:

The partial class is generated if either the user-provided interface or the user-provided partial class is present, or both.

The "zipping" key in my case is the name of the generated partial class.

#### Question:

How to currently achieve this (in a least cache-unfriendly way)?

**Comment by 333fred** [comment]
If you need to search among all values in a provider for a key, then I don't think there's a cache-friendly way to achieve it. You'll have to Combine and then pull back out via SelectMany.

**Reply by n9** [reply]
I currently use two different incremental generators (one for the interface and the other for the partial class). The problem is with a shared state. I am considering creating a shared state with a third one.
